volume: Change owner of symlinks too

mauriciovasquezbernal opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This commit uses Lchown instead of Chown to change the owner of symlinks too.
It doesn't change any behaviour. However, it could avoid some confusions as the
symlinks are updated to the correct owner too.

Please consider the following Pod with a configMap:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: myconfigmap
  namespace: default
data:
  content: '"This is a ConfigMap"'
---
apiVersion: v1
kind: Pod
metadata:
  name: mypod
spec:
  securityContext:
    fsGroup: 2000
  containers:
  - name: container1
    image: busybox
    command: ["sh"]
    args: ["-c", "sleep infinity"]
    volumeMounts:
    - name: config-volume
      mountPath: /etc/configmap/content
  volumes:
  - name: config-volume
    configMap:
      name: myconfigmap
      defaultMode: 0400
```

The current behaviour shows that the group owner is `0` for the symlink point to the real content file:
```
$ kubectl exec -it mypod -- /bin/sh -c 'ls -la /etc/configmap/content/content'
lrwxrwxrwx    1 0     0            14 Sep 18 15:15 /etc/configmap/content/content -> ..data/content
```

The behaviour after this PR shows as group owner `2000` as specified in `fsGroup`:
```
kubectl exec -it mypod -- /bin/sh -c 'ls -la /etc/configmap/content/content'
lrwxrwxrwx    1 0     2000            14 Sep 18 15:15 /etc/configmap/content/content -> ..data/content
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mauriciovasquezbernal committed:
volume: Change owner of symlinks too

This commit uses Lchown instead of Chown to change the owner of symlinks too.
It doesn't change any behaviour. However, it could avoid some confusions as the
symlinks are updated to the correct owner too.

mauriciovasquezbernal commented:
/sig storage 
/release-note-none

gnufied commented on pkg/volume/volume_linux.go:
```diff
@@ -82,32 +82,22 @@ func legacyOwnershipChange(mounter Mounter, fsGroup *int64) error {
-	if err != nil {
-		klog.Errorf("Chown failed on %v: %v", filename, err)
-	}
-
```
Isn't this changing the existing behaviour?

mauriciovasquezbernal replied:
Strictly speaking, yes. Now the group owner of the symlinks is updated too. Am I missing anything else?

gnufied commented:
/assign @sjenning @gnufied

gnufied commented:
/ok-to-test

jingxu97 commented:
/lgtm

jingxu97 commented:
could you add unit test? Thanks!

cc @msau42

mauriciovasquezbernal committed:
volume: Add unit test for SetVolumeOwnership owners

Former TestSetVolumeOwnership only checks the mode of the files. This commit
adds a new TestSetVolumeOwnershipOwner that checks the ownership of the files.

mauriciovasquezbernal commented:
@jingxu97 I added the unit test.

mauriciovasquezbernal commented:
/retest

mauriciovasquezbernal commented:
@gnufied @sjenning any chance you can give it a look?

gnufied commented:
It looks good to me as well. But it would be nice to have a confirmation from @sjenning because he originally made the change that prevented permissions of symlinks from being changed. But at that time - we were calling plain` chown` which evaluates the symlink and now you are using `lchown`.

joelsmith commented:
Nice that you were able to get rid of an extra stat syscall by using -1 for the UID!

/lgtm

sjenning commented:
/approve

gnufied commented:
/approve
